ddht/v5_1/client.py

```diff
@@ -396,6 +399,82 @@ async def find_nodes(
 
                 return responses
 
+    @asynccontextmanager
+    async def stream_find_nodes(
+        self,
+        node_id: NodeID,
+        endpoint: Endpoint,
+        distances: Collection[int],
+        *,
+        request_id: Optional[bytes] = None,
+    ) -> AsyncIterator[trio.abc.ReceiveChannel[InboundMessage[FoundNodesMessage]]]:
+        with self.request_tracker.reserve_request_id(node_id, request_id) as request_id:
+            request = AnyOutboundMessage(
+                FindNodeMessage(request_id, tuple(distances)), endpoint, node_id,
+            )
+
+            send_channel, receive_channel = trio.open_memory_channel[
+                InboundMessage[FoundNodesMessage]
+            ](256)
+            async with trio.open_nursery() as nursery:
+                nursery.start_soon(
+                    self._stream_find_nodes_response,
+                    request,
+                    request_id,
+                    FoundNodesMessage,
+                    send_channel,
+                )
+
+                try:
+                    async with receive_channel:
+                        try:
+                            yield receive_channel
+                        except trio.EndOfChannel as err:
+                            raise trio.TooSlowError from err
+                finally:
+                    nursery.cancel_scope.cancel()
+
+    async def _stream_find_nodes_response(
+        self,
+        request: AnyOutboundMessage,
+        request_id: Optional[bytes],
+        response_message_type: Type[BaseMessage],
+        send_channel: trio.abc.SendChannel[InboundMessage[FoundNodesMessage]],
+    ) -> None:
+        num_responses = 0
+
+        with trio.move_on_after(REQUEST_RESPONSE_TIMEOUT) as scope:
+            subscription_ctx = self.dispatcher.subscription_manager.subscribe(
+                response_message_type,
+                request.receiver_endpoint,
+                request.receiver_node_id,
+            )
+            async with subscription_ctx as subscription:
+                await self.dispatcher.send_message(request)
+
+                async with send_channel:
+                    async for response in subscription:
+                        if response.message.total == 0:
+                            raise ValidationError(
+                                f"Invalid `total` counter in response: total={response.message.total}"
+                            )
+
+                        if response.request_id != request_id:
+                            continue
+                        else:
+                            num_responses += 1
+                            await send_channel.send(response)
```

Comment:
I think this can raise `trio.BrokenResourceError` if the user exits the context before all responses have been returned.  You should be able to test this by setting up a request response where the response has multiple responses, but the receiver breaks out of the context after receiving the first response.